File: Python/trex_selector_methods/trex/demo_trex_05_mmap.py
# ==============================================================================
# demo_trex_05_mmap.py
# ==============================================================================
#
# Classical T-Rex Selector demo — memory-mapped workflows.
# Mirrors R/trex_selector_methods/trex/demo_trex_05_mmap.R and
# cpp/trex_selector_methods/trex/demo_trex_05_mmap.cpp.
#
# Demo content:
#
#  Part A: In-memory X + use_memory_mapping=True.
#          Setting use_memory_mapping=True activates both:
#            (1) Memory-mapped storage for internal dummy matrices D.
#            (2) Solver LARS-path serialization to disk between T-loop iterations.
#          X itself stays fully in RAM.
#          Exact mirror of C++ Demo A.
#
#  Part B: Fully memory-mapped pipeline.
#          X is written to a temporary binary file via numpy_to_memmap().
#          X_mmap.to_numpy() (a zero-copy OS-paged view) is passed to
#          TRexSelector; use_memory_mapping=True adds D-mmap + solver
#          serialization on top.
#          Mirrors C++ Demo B (SyntheticDataMapped) and R Part B
#          (execute_with_temp_mmap).
#
# Seed note:
#   Both parts use seed=58 for DGP and TRexSelector (single-run reproducibility,
#   mirrors R Part A/B).  This is intentional for a single-run demo; MC demos
#   use seed=-1 (hardware entropy) for valid FDR estimates.
#
# ==============================================================================

import sys
import os

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
if _THIS_DIR not in sys.path:
    sys.path.insert(0, _THIS_DIR)

import tempfile
import numpy as np
import trex_selector as ts
from trex_selector.utils import numpy_to_memmap

from dgp_gauss_snr import dgp_gauss_snr
from trex_sim_common import _make_trex_ctrl_from_dict
from trex_helpers import print_single_run_result, save_selection_csv

# ==============================================================================
# Shared demo configuration
# ==============================================================================
# Mirrors make_mmap_demo_config() in C++ and DEMO_CFG in R.
#
# Fixed true support (0-based) — note: different from demo_trex_01.
# C++ (0-based): {27, 149, 398, 420, 4}
# R   (1-based): {28, 150, 399, 421, 5}
_TRUE_SUPPORT = np.array([27, 149, 398, 420, 4], dtype=np.intp)
_TRUE_COEFS   = np.ones(5, dtype=np.float64)     # rnd_coef = False
_SNR          = 1.0
_TFDR         = 0.1
_SEED         = 58

_OUT_DIR = os.path.join(_THIS_DIR, "simulation_results")

# Shared control: mirrors make_mmap_trex_ctrl() in C++ and R
_MMAP_CTRL_DICT = dict(
    solver_name               = "TLARS",
    K                         = 20,
    max_dummy_multiplier      = 10,
    use_max_T_stop            = True,
    lloop_strategy            = "HCONCAT",
    tloop_stagnation_stop     = True,
    use_memory_mapping        = True,   # activates D-mmap + solver serialization
    lambda2                   = 0.1,
    rho_afs                   = 0.3,
    ncgmp_variant             = 0,
    parallel_rnd_experiments  = False,  # mirrors R use_openmp=FALSE
)


# ==============================================================================
# Shared print helper (mirrors R .print_and_save_d02)
# ==============================================================================

def _print_and_save(stem, dat, sel, tFDR):
    from trex_sim_common import compute_fdp, compute_tpp
    true_support = dat["true_support"]
    selected     = list(sel.selected_indices)
    n_sel        = len(selected)
    true_set     = set(true_support.tolist())
    n_tp         = len(set(selected) & true_set)
    n_fp         = n_sel - n_tp
    fdp_val      = compute_fdp(selected, true_support.tolist())
    tpp_val      = compute_tpp(selected, true_support.tolist())

    print("-" * 70)
    print(f"  Selected indices (0-based): {{{', '.join(str(x) for x in sorted(selected))}}}")
    print(f"  FDP = {fdp_val:.4f}  |  TPP = {tpp_val:.4f}  (target tFDR <= {tFDR:.2f})")
    print(f"  T_stop = {sel.T_stop},  L = {sel.L}")
    print("-" * 70 + "\n")

    save_selection_csv(
        _OUT_DIR, stem,
        p=dat["p"],
        phi_prime=sel.phi_prime,
        selected_0based=selected,
        true_support_0based=true_support.tolist(),
    )


# ==============================================================================
# Part A: In-memory X + use_memory_mapping=True
# ==============================================================================

def part_a_d_mmap_solver_serial():
    """
    Mirrors C++ demo_TRexSelector_d_mmap_solver_serial() and R trx_02_part_a().
    """
    def run_single(high_dim):
        n = 1000 if high_dim else 5000
        p = 5000 if high_dim else 1000
        dim_label = "High-dimensional (p > n)" if high_dim else "Low-dimensional (n > p)"

        print(f"\n{'=' * 70}")
        print(f"Part A: In-Memory X  +  use_memory_mapping=True")
        print(f"  Activates D-mmap and solver serialization inside TRexSelector.")
        print(f"  {dim_label}  (n = {n}, p = {p})\n")

        print("Generating synthetic data (in-memory) ...")
        dat = dgp_gauss_snr(n, p, _TRUE_SUPPORT, coefs=_TRUE_COEFS,
                            snr=_SNR, seed=_SEED)

        ctrl = _make_trex_ctrl_from_dict(_MMAP_CTRL_DICT)

        print("Running T-Rex Selector ...\n")
        sel = ts.TRexSelector(
            np.asfortranarray(dat["X"]),
            dat["y"],
            tFDR=_TFDR,
            trex_control=ctrl,
            seed=_SEED,
            verbose=True,
        )
        sel.select()

        stem = f"d02_trex_mmap_demo_a_n{n}_p{p}"
        _print_and_save(stem, dat, sel, _TFDR)

    run_single(high_dim=False)
    run_single(high_dim=True)


# ==============================================================================
# Part B: Memory-mapped X + use_memory_mapping=True
# ==============================================================================

def part_b_full_mmap():
    """
    Mirrors C++ demo_TRexSelector_full_mmap() and R trx_02_part_b().

    X is written to a temporary binary file; X_mmap.to_numpy() (a zero-copy
    Fortran-contiguous float64 view backed by the mmap buffer) is passed to
    TRexSelector.  The temp file is removed in a finally block — exception-safe,
    mirrors C++ RAII MmapFileGuard and R execute_with_temp_mmap / on.exit().
    """
    def run_single(high_dim):
        n = 1000 if high_dim else 5000
        p = 5000 if high_dim else 1000
        dim_label = "High-dimensional (p > n)" if high_dim else "Low-dimensional (n > p)"

        print(f"\n{'=' * 70}")
        print(f"Part B: Memory-Mapped X  +  use_memory_mapping=True")
        print(f"  Full mmap pipeline: X mmap + D mmap + solver serialization.")
        print(f"  {dim_label}  (n = {n}, p = {p})\n")

        print("Generating synthetic data ...")
        dat = dgp_gauss_snr(n, p, _TRUE_SUPPORT, coefs=_TRUE_COEFS,
                            snr=_SNR, seed=_SEED)

        ctrl = _make_trex_ctrl_from_dict(_MMAP_CTRL_DICT)

        # Create temporary backing file.  os.close(fd) releases the OS file
        # descriptor; numpy_to_memmap re-opens the path for writing.
        fd, mmap_path = tempfile.mkstemp(suffix=".bin")
        os.close(fd)

        try:
            print(f"Converting X to memory-mapped file ({mmap_path}) ...")
            X_mmap = numpy_to_memmap(mmap_path, dat["X"])

            print("Running T-Rex Selector ...\n")
            # X_mmap.to_numpy() → Fortran-contiguous float64 mmap view.
            # TRexSelector.__init__ calls np.asfortranarray(X) which returns
            # the view unchanged (already F-contiguous float64, zero copy).
            sel = ts.TRexSelector(
                X_mmap.to_numpy(),
                dat["y"],
                tFDR=_TFDR,
                trex_control=ctrl,
                seed=_SEED,
                verbose=True,
            )
            sel.select()
        finally:
            # Exception-safe cleanup.  On CPython the MemoryMappedMatrix is
            # reference-counted; it goes out of scope (and releases the mmap
            # mapping) before this finally block runs.
            try:
                os.unlink(mmap_path)
                print(f"[Info] Temp mmap file removed: {mmap_path}")
            except OSError:
                pass

        stem = f"d02_trex_mmap_demo_b_n{n}_p{p}"
        _print_and_save(stem, dat, sel, _TFDR)

    run_single(high_dim=False)
    run_single(high_dim=True)


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    part_a_d_mmap_solver_serial()
    part_b_full_mmap()
    print("\ndemo_trex_05_mmap.py complete.")

File: R/trex_selector_methods/trex/demo_trex_05_mmap/demo_trex_05_mmap.R
# ==============================================================================
# demo_trex_05_mmap.R
# ==============================================================================
#
# Classical T-Rex selector demo.
# Part 5 demonstrates memory-mapped usage patterns in the R package TRexSelectorNeo.
#
# The file mirrors:
# cpp/trex_selector_methods/trex/demo_trex_05_mmap.cpp
#
# Demo content:
# ----------------------------------------------------------------
#
#  Part A: In-memory X + use_memory_mapping = TRUE.
#          Enables internal D-mmap and solver serialization within
#          TRexSelector (via trex_control(use_memory_mapping = TRUE)).
#          Exact mirror of C++ Demo A.
#
#
#  Part B: Fully memory-mapped X + use_memory_mapping = TRUE.
#          In C++ this uses SyntheticDataMapped (X never fully in RAM).
#          The R package TRexSelectorNeo supports MemoryMappedMatrix objects
#          for X (via convert_to_memory_mapped / mmap_matrix), but creating
#          a memory-mapped version of a generated matrix requires writing it
#          to disk first.
#          This part mirrors the statistical scenario
#          (same n, p, support, snr, ctrl) while demonstrating the R mmap
#          interface for completeness.
#
# ==============================================================================

library(TRexSelectorNeo)

# ==============================================================================
# Setup
# ==============================================================================

this_dir_ <- tryCatch(
  dirname(normalizePath(sys.frame(1L)$ofile)),
  error = function(e) {
    args <- commandArgs(trailingOnly = FALSE)
    file_arg <- grep("--file=", args, value = TRUE)
    if (length(file_arg) > 0) dirname(normalizePath(sub("--file=", "", file_arg[1]))) else "."
  }
)
source(file.path(this_dir_, "..", "simulation_utils.R"))
source(file.path(this_dir_, "trex_sim_utils.R"))


# ==============================================================================
# Global Simulation Parameters
# ==============================================================================

OUT_DIR <- file.path(this_dir_, "simulation_results")
dir.create(OUT_DIR, recursive = TRUE, showWarnings = FALSE)

# Run flags
run_part_a <- TRUE
run_part_b <- TRUE

# Shared demo config (mirrors make_mmap_demo_config() in C++)
# C++ true_support (0-based): {27, 149, 398, 420, 4}
# R (1-based): {28, 150, 399, 421, 5}
DEMO_CFG <- list(
  true_support = c(28L, 150L, 399L, 421L, 5L),
  true_coefs   = c(1, 1, 1, 1, 1),   # rnd_coef = FALSE
  snr          = 1.0,
  tFDR         = 0.1
)


# ==============================================================================
# Shared helper
# ==============================================================================

#' Print and save single-run results.
.print_and_save_d02 <- function(stem, dat, selector, tFDR) {
  true_support <- dat$true_support
  selected     <- selector$selected_indices
  n_sel        <- length(selected)
  n_tp         <- length(intersect(selected, true_support))
  n_fp         <- n_sel - n_tp
  tpp          <- TRexSelectorNeo::compute_tpp(selected, true_support)
  fdp          <- TRexSelectorNeo::compute_fdp(selected, true_support)

  cat(strrep("-", 70L), "\n")
  cat(sprintf("  Selected indices: {%s}\n", paste(selected, collapse = ", ")))
  cat(sprintf("  FDP = %.4f  |  TPP = %.4f  (target tFDR <= %.2f)\n",
              fdp, tpp, tFDR))
  cat(sprintf("  T_stop = %d,  L = %d\n", selector$T_stop, selector$L))
  cat(strrep("-", 70L), "\n\n")

  # Save per-variable CSV (mirrors save_selection_csv() in C++)
  p          <- ncol(dat$X)
  phi_prime  <- selector$phi_prime
  is_sel     <- seq_len(p) %in% selected
  is_true    <- seq_len(p) %in% true_support
  csv_df     <- data.frame(
    variable_index = seq_len(p) - 1L,   # 0-based to match C++ output
    phi_prime      = round(phi_prime, 6L),
    selected       = as.integer(is_sel),
    true_positive  = as.integer(is_true)
  )
  csv_path <- file.path(OUT_DIR, paste0(stem, ".csv"))
  utils::write.csv(csv_df, csv_path, row.names = FALSE, quote = FALSE)
  cat(sprintf("[Info] CSV saved to: %s\n\n", csv_path))
}


# ==============================================================================
# Part A: In-memory X + use_memory_mapping = TRUE
# ==============================================================================
# Mirrors demo_TRexSelector_d_mmap_solver_serial() in C++.

trex_05_part_a <- function() {

  run_single <- function(high_dim) {
    cfg       <- DEMO_CFG
    cfg$n     <- if (high_dim) 1000L else 5000L
    cfg$p     <- if (high_dim) 5000L else 1000L

    .print_demo_header(
      "Part A: In-Memory X  +  use_memory_mapping = TRUE",
      "Activates D-mmap and solver serialization inside TRexSelector.",
      high_dim, cfg$n, cfg$p
    )

    ctrl <- make_mmap_trex_ctrl(solver = "TLARS")

    cat("Generating synthetic data (in-memory) ...\n")
    dat <- dgp_gauss_snr(cfg$n, cfg$p, cfg$true_support, coefs = cfg$true_coefs,
                         snr = cfg$snr, seed = 58L)

    cat("Running T-Rex Selector ...\n\n")
    selector <- TRexSelector$new(
      X       = dat$X,
      y       = dat$y,
      tFDR    = cfg$tFDR,
      seed    = 58L, # deterministic seed for reproducibility
      verbose = TRUE,
      control = ctrl
    )
    selector$select()

    stem <- sprintf("d02_trex_mmap_demo_a_n%d_p%d", cfg$n, cfg$p)
    .print_and_save_d02(stem, dat, selector, cfg$tFDR)
  }

  run_single(high_dim = FALSE)
  run_single(high_dim = TRUE)
}


# ==============================================================================
# Part B: Memory-mapped X + use_memory_mapping = TRUE
# ==============================================================================
# In C++: uses SyntheticDataMapped so X is never fully in RAM.
# In R:   demonstrates the TRexSelector MemoryMappedMatrix interface by
#         writing X to a temporary mmap file before passing it to the selector.
#         use_memory_mapping = TRUE adds D-mmap + solver serialization on top.

trex_05_part_b <- function() {

  run_single <- function(high_dim) {
    cfg       <- DEMO_CFG
    cfg$n     <- if (high_dim) 1000L else 5000L
    cfg$p     <- if (high_dim) 5000L else 1000L

    .print_demo_header(
      "Part B: Memory-Mapped X  +  use_memory_mapping = TRUE",
      "Full mmap pipeline: X mmap + D mmap + solver serialization.",
      high_dim, cfg$n, cfg$p
    )

    ctrl <- make_mmap_trex_ctrl(solver = "TLARS")

    # Generate data in-memory, then convert X to a memory-mapped file.
    cat("Generating synthetic data ...\n")
    dat <- dgp_gauss_snr(cfg$n, cfg$p, cfg$true_support, coefs = cfg$true_coefs,
                         snr = cfg$snr, seed = 58L)

    stem <- sprintf("d02_trex_mmap_demo_b_n%d_p%d", cfg$n, cfg$p)
    cat("Converting X to memory-mapped file and running T-Rex Selector...\n\n")

    execute_with_temp_mmap(dat$X, function(X_mmap) {
      selector <- TRexSelector$new(
        X       = X_mmap,
        y       = dat$y,
        tFDR    = cfg$tFDR,
        seed    = 58L, # deterministic seed for reproducibility
        verbose = TRUE,
        control = ctrl
      )
      selector$select()
      .print_and_save_d02(stem, dat, selector, cfg$tFDR)
    })
  }

  run_single(high_dim = FALSE)
  run_single(high_dim = TRUE)
}


# ==============================================================================
# Main
# -------------------
# Run parts A and B
# ==============================================================================

if (run_part_a) trex_05_part_a()
if (run_part_b) trex_05_part_b()
cat("\ndemo_trex_05_mmap.R complete.\n")
